### ISSUE TITLE
The compiler should be warned when some misleading code is being used

### ISSUE THREAD

**Issue by 3442853561** [Manishearth/rust-clippy#1308]
The compiler should be warned when some misleading code is being used
for example:
- Use && and || at the same time without parentheses. This may cause programmers to misunderstand their priorities.
- The identifiers are case-sensitive only. Programmers may mix them.
- Identifiers are defined in capitalized form for keywords. Programmers may mix them.
- Use return in the last sentence of the function. Just because it is not beautiful.

**Comment by llogiq** [comment]
Thanks for your interest in improving clippy. To comment on your points one by one:

* I believe the precedence of `&&` over `||` is widely understood
* rustc already has a lint that guards following the naming conventions
* could you specify an example that is within the naming conventions and causes any readability problem?
* clippy has had `needless_return` for quite a long time now

Perhaps I misunderstood you (English is not my mother tongue), if so, I welcome further clarification. All in all, if you can show that the naming conventions lint has an oversight, we can either extend rustc or patch that 'hole' with a clippy lint. I think that would be a valuable endeavor.

**Comment by 3442853561** [comment]
@llogiq Thank you for reply. If you are a Chinese developer you will understand `&&` greater than `||` isn't clear. This is something that we have been neglecting in the process of learning to program. After all, Professor Tan ,who is the author of the official textbooks, always overlooks some of the key points. So, the Chinese developers form a rule that must use parentheses to denote the precedence of an operation, except the `+` , `-` , `*` and `/` these four already very clear operation.

**Comment by 3442853561** [comment]
@llogiq 

> Use return in the last sentence of the function. Just because it is not beautiful.

 I mean the use of `return` is not beautiful, but some developers, who was C developer, always forget this.

**Comment by 3442853561** [comment]
some bad show:

example for case-sensitive only:

```
let foo = 5;
let FOO = 2;
```

example for defined in capitalized form for keywords

```
let IF = 0；
let LOOP = 1;
let RETURN = 2;
fn FOR() {
    ...
}
```

**Comment by llogiq** [comment]
A lint requiring parenthesis on mixed use of || and && is of little value outside of China, so it would probably be `allow` by default.

As I stated, the needless_return lint will already catch returning values as in your example, and has done so for some time.

The naming convention lints built into rustc catches all your examples already, so adding yet another lint for the same thing is of little value.

**Comment by mcarton** [comment]
> A lint requiring parenthesis on mixed use of || and && is of little value outside of China, so it would probably be `allow` by default.

[Some people like it](https://github.com/rust-lang-nursery/fmt-rfcs/issues/18#issuecomment-252302195)

For the rest, I agree it already exists.

**Comment by llogiq** [comment]
In that case, this is probably better handled by rustfmt than clippy, because the latter will likely drown out most other lint on any codebase doing even moderately complex predicates.
